include/internal/refcount.h

```diff
@@ -44,106 +45,104 @@ static inline int CRYPTO_UP_REF(_Atomic int *val, int *ret,
  * to mutable members doesn't have to be serialized anymore, which would
  * otherwise imply an acquire fence. Hence conditional acquire fence...
  */
-static inline int CRYPTO_DOWN_REF(_Atomic int *val, int *ret,
-                                  ossl_unused void *lock)
+static inline int CRYPTO_DOWN_REF(CRYPTO_REF_COUNT *refcnt, int *ret)
 {
-    *ret = atomic_fetch_sub_explicit(val, 1, memory_order_relaxed) - 1;
+    *ret = atomic_fetch_sub_explicit(&refcnt->val, 1, memory_order_relaxed) - 1;
     if (*ret == 0)
         atomic_thread_fence(memory_order_acquire);
     return 1;
 }
 
-static inline int CRYPTO_GET_REF(_Atomic int *val, int *ret,
-                                 ossl_unused void *lock)
+static inline int CRYPTO_GET_REF(CRYPTO_REF_COUNT *refcnt, int *ret)
```

Comment:
This function does not belong in this file. Functions using this is probably yet an other abuse of the API.

Reply:
Why?  Getting a reference count seems like a natural part of reference counting.

I agree that the API is a bit rubbish.

Reply:
Let me rephrase that: This function should not exist.

CRYPTO_UP_REF() should not have returned a number, CRYPTO_DOWN_REF should only indicate that the reference count has become 0 and that you should free it.

CRYPTO_GET_REF(), specially with relaxed semantics, will get you the wrong answer.

Reply:
It's only used in one place `ossl_quic_attach_stream` checking for user error.  This function was supposed to be removed or made internal only if I remember correctly.  @hlandau?  In either case it doesn't need to exist.

Reply:
Raised as  #21278